src/file.cpp

```diff
@@ -974,20 +968,28 @@ namespace libtorrent
 	{
 #if defined TORRENT_WINDOWS && !defined TORRENT_MINGW
 #if TORRENT_USE_WSTRING
-		wchar_t* cwd { ::_wgetcwd(nullptr, 0) };
-		std::shared_ptr<wchar_t> holder(cwd, &std::free);
-		std::string ret { cwd == nullptr ? "" : convert_from_wstring(cwd) };
+#define GetCurrentDir_ ::_wgetcwd
+#define PathCharType_ wchar_t
+#define ConvertFrom_ convert_from_wstring
 #else
-		char* cwd { ::_getcwd(nullptr, 0) };
-		std::shared_ptr<char> holder(cwd, &std::free);
-		std::string ret { cwd == nullptr ? "" : convert_from_native(cwd) };
+#define GetCurrentDir_ ::_getcwd
+#define PathCharType_ char
+#define ConvertFrom_ convert_from_native
 #endif // TORRENT_USE_WSTRING
 #else
-		char* cwd { ::getcwd(nullptr, 0) };
-		std::shared_ptr<char> holder(cwd, &std::free);
-		std::string ret { cwd == nullptr ? "/" : convert_from_native(cwd) };
-#endif
+#define GetCurrentDir_ ::getcwd
+#define PathCharType_ char
+#define ConvertFrom_ convert_from_native
+#endif
+		PathCharType_* cwd { GetCurrentDir_(nullptr, 0) };
+		if (cwd == nullptr)
+			aux::throw_ex<system_error>(error_code(errno, generic_category()));
+		std::shared_ptr<PathCharType_> holder(cwd, &std::free);
```

Comment:
for this you could add a function like this: ``template <typename T> std::unique_ptr<T, decltype(&std::free)> make_free_holder(T* ptr) { return std::unique_ptr<T, decltype(&std::free)>(ptr, &std::free); }``.

With such function, you could say: ``auto holder = make_free_holder(cwd);``, without having to explicitly mention the pointer type